qualtran/bloqs/arithmetic/permutation.py

```diff
@@ -0,0 +1,248 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import cast, Sequence, Set, TYPE_CHECKING, TypeAlias, Union
+
+from attrs import field, frozen
+
+from qualtran import (
+    Bloq,
+    bloq_example,
+    BoundedQUInt,
+    DecomposeTypeError,
+    QBit,
+    QUInt,
+    Signature,
+    Soquet,
+)
+from qualtran.bloqs.arithmetic.addition import XorK
+from qualtran.bloqs.arithmetic.comparison import EqualsAConstant
+from qualtran.linalg.permutation import (
+    CycleT,
+    decompose_permutation_into_cycles,
+    decompose_sparse_prefix_permutation_into_cycles,
+)
+from qualtran.symbolics import bit_length, is_symbolic, Shaped, slen, SymbolicInt
+
+if TYPE_CHECKING:
+    import sympy
+
+    from qualtran import BloqBuilder, SoquetT
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+SymbolicCycleT: TypeAlias = Union[CycleT, Shaped]
+
+
+def _convert_cycle(cycle) -> Union[tuple[int, ...], Shaped]:
+    if isinstance(cycle, Shaped):
+        return cycle
+    return tuple(cycle)
+
+
+@frozen
+class PermutationCycle(Bloq):
+    r"""Apply a single permutation cycle on the basis states.
+
+    Args:
+        N: the total size the permutation acts on.
+        cycle: the permutation cycle to apply.
+
+    Registers:
+        q: integer register storing a value in [0, ..., N - 1]
```

Comment:
lets reserve the register name "q" for a single qubit. We generally use "x" (my pref) or "a" for an arbitrary value